core/dbt/graph/selector.py

```diff
@@ -258,8 +258,25 @@ def get_selected(self, spec: SelectionSpec) -> Set[UniqueId]:
                 - selectors can filter the nodes after all of them have been
                   selected
         """
-        selected_nodes = self.select_nodes(spec)
+        selected_nodes, indirect_only = self.select_nodes(spec)
         filtered_nodes = self.filter_selection(selected_nodes)
+
+        if indirect_only:
+            filtered_unused_nodes = self.filter_selection(indirect_only)
+            # log anything that didn't make the cut
+            if filtered_unused_nodes:
+                unused_node_names = []
+                for unique_id in filtered_unused_nodes:
+                    name = self.manifest.nodes[unique_id].name
+                    unused_node_names.append(name)
+                unused_nodes_str = ("\n  - ").join(unused_node_names)
+                msg = (
+                    f"\nSome resources were excluded because at least one parent is missing:"
+                    f"\n  - {unused_nodes_str}"
+                    f"\nUse the --greedy flag to include them"
```

Comment:
I'm on the fence between:
- it's worthwhile surfacing enough info to make clear what's happening
- taking up prime real estate with a list of potentially dozens/hundreds of not-greedily-included tests

What do you think about splitting this into two log messages:
- [info level] "Some resources were excluded because at least one parent is ~missing~ not selected. Use the --greedy flag to include them."
- [debug level] List every specific resource that was excluded because it had at least one parent unseelcted

Also - as I typed this, I realized "missing" may imply "missing from the database," when really we mean "not currently selected"

Reply:
Good shout on _missing_ vs _not selected_.

How about listing the first, say, 3 at info level and the whole lot at debug:
```
Some resources were excluded because at least one parent is not selected:
- excluded_test_1
- excluded_test_2
- excluded_test_3
- and 4 more
Use the --greedy flag to include them
```

My main concern is balancing the needs of someone who's brand new and doesn't know where to go for debug-level messages (esp in the IDE), while not annoying long-time power users.

Reply:
On the note of naming, will the "resources" always be tests? I'm assuming so since the `--greedy` flag is only available in the `dbt test` context. In which case could we give it a more concrete name?

Reply:
I'm happy with the balanced approach: listing the first three "and X more." What do you think about making the message inline rather than a list?

Today, the only resources that can be selected indirectly are tests; that's stated [here](https://github.com/dbt-labs/dbt/blob/df7ec3fb37396e67b7986cea99a370bba7c54ca0/core/dbt/graph/selector.py#L33-L42). The `--greedy` flag should be available to both `dbt test` and `dbt build`, but in both cases, the only things that show up in this list will be tests.

We did implement this in such a way as to minimize hard-coding that this is for tests only; while I can't think of one right now, we might eventually extend this to other resource types, though. But I'm happy to hard-code "tests" in the log message, to avoid user confusion, and update it if/when we widen the net of indirect selection.

Reply:
>What do you think about making the message inline rather than a list?

I just gave it a go, but found it very hard to visually parse when they all ran together instead of being newline delimited. Unless you feel really strongly that it should be all together, I'd rather not. 

Have hardcoded tests into the message for now 👍